pkg/ccl/changefeedccl/sink_cloudstorage.go

```diff
@@ -648,21 +662,55 @@ func (s *cloudStorageSink) flushFile(ctx context.Context, file *cloudStorageSink
 		return file.flushToStorage(ctx, s.es, dest, s.metrics)
 	}
 
-	s.flushGroup.Add(1)
-	go func() {
-		defer s.flushGroup.Done()
-		// NB: must use s.flushCtx; ctx may be short lived (i.e. cancelled).
-		if err := file.flushToStorage(s.flushCtx, s.es, dest, s.metrics); err != nil {
-			log.Errorf(ctx, "error flushing file to storage: %s", err)
-			// We must use the same type for error we store in flushErr.
-			s.flushErr.CompareAndSwap(nil, &flushError{error: err})
+	s.flushQueue.Lock()
+	defer s.flushQueue.Unlock()
+	if s.flushQueue.q == nil {
+		// Start a new go routine if the flush queue is empty.
+		s.flushGroup.Add(1)
+		go func() {
+			s.asyncFlush(file, dest)
+		}()
+	} else {
+		// When queue is not empty, queue the file to be flushed by active
+		// go routine.
+		if logQueueDepth.ShouldLog() {
+			depth := len(s.flushQueue.q)
+			log.Infof(ctx, "changefeed flush queue depth %d; ~%d bytes to flush",
+				depth, int64(depth)*s.targetMaxFileSize)
 		}
-	}()
+		s.flushQueue.q = append(s.flushQueue.q, destFile{file: file, dest: dest})
+	}
 	return nil
 }
 
-type flushError struct {
-	error
+func (s *cloudStorageSink) asyncFlush(file *cloudStorageSinkFile, dest string) {
+	defer s.flushGroup.Done()
+
+	pop := func(err error) destFile {
+		s.flushQueue.Lock()
+		defer s.flushQueue.Unlock()
+		s.flushQueue.err = err
```

Comment:
check to make sure this isn't populated already? It seems to me like this is racy because you don't check for an error when you go to append to the queue. Imagine the case where you have one flush call (and so nothing in the queue) and then it gets an error and assigns here. The goroutine returns but another call to flush above launches a new goroutine which then succeeds and clobbers this error. 

I feel like you should just simplify this code so that the file currently being flushed is at the front of the queue and it gets popped after it has succeeded. I think it'll simplify some things.

Reply:
Good catch.
The reality is that I used to have exactly that code, but undoing a more complex version, I left the complexity I should have had.